arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp

```diff
@@ -0,0 +1,1264 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2020 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Heiko Kernbach
+/// @author Lars Maier
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+
+#include <Basics/VelocyPackHelper.h>
+#include <velocypack/Collection.h>
+#include <velocypack/Iterator.h>
+#include <velocypack/velocypack-aliases.h>
+
+#include <iostream>
+#include <list>
+
+#include "Extractor.h"
+#include "Interpreter.h"
+#include "Primitives.h"
+
+namespace arangodb::greenspun {
+
+EvalResult Prim_Min(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  bool set = false;
+  auto tmp = double{0};
+  for (auto p : VPackArrayIterator(params)) {
+    if (p.isNumber<double>()) {
+      if (!set) {
+        tmp = p.getNumericValue<double>();
+        set = true;
+      } else {
+        tmp = std::min(tmp, p.getNumericValue<double>());
+      }
+    } else {
+      return EvalError("expected double, found: " + p.toJson());
+    }
+  }
+  if (set) {
+    result.add(VPackValue(tmp));
+  } else {
+    result.add(VPackSlice::noneSlice());
+  }
+  return {};
+}
+
+EvalResult Prim_Max(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  bool set = false;
+  auto tmp = double{0};
+  for (auto p : VPackArrayIterator(params)) {
+    if (p.isNumber<double>()) {
+      if (!set) {
+        tmp = p.getNumericValue<double>();
+        set = true;
+      } else {
+        tmp = std::max(tmp, p.getNumericValue<double>());
+      }
+    } else {
+      return EvalError("expected double, found: " + p.toJson());
+    }
+  }
+  if (set) {
+    result.add(VPackValue(tmp));
+  } else {
+    result.add(VPackSlice::noneSlice());
+  }
+  return {};
+}
+
+EvalResult Prim_Avg(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  auto tmp = double{0};
+  for (auto p : VPackArrayIterator(params)) {
+    if (p.isNumber<double>()) {
+      tmp += p.getNumericValue<double>();
+    } else {
+      return EvalError("expected double, found: " + p.toJson());
+    }
+  }
+  auto length = params.length();
+  result.add(VPackValue(length == 0 ? tmp : tmp / length));
```

Comment:
Is the average of an empty array `0`, or should it rather be something like null?

Reply:
In math(s) the empty sum is 0 and thus the empty average is 0.